feat(agent): add container list handler

johnstcn opened the pull request:
Fixes https://github.com/coder/coder/issues/16268

This PR adds an API endpoint to coderd `/api/v2/workspaceagents/:id/containers` that allows listing containers visible to the agent. Optional filtering by labels is supported.

A later PR will consume this endpoint and use it to expose this information to users.

Under the hood, this operates in a very similar manner to the `listening-ports` endpoint -- `coderd` dials the workspace agent and hits an agent endpoint exposed over `tailnet`. 

This initial implementation only supports listing containers using the Docker CLI. Support for other data sources may be added at a future date.

Sample response:

```
{
  "containers": [
    {
      "created_at": "2025-02-05T12:53:18.943217568Z",
      "id": "1f119acf43202781f7699534fe3613ce2764369148738a16970fb124233a7c58",
      "name": "affectionate_black",
      "image": "busybox:latest",
      "labels": {

      },
      "running": false,
      "ports": [],
      "status": "exited with code 127: failed to create task for container: failed to create shim task: OCI runtime create failed: runc create failed: unable to start container process: exec: \"doesnotexist\": executable file not found in $PATH: unknown",
      "volumes": {

      }
    },
    {
      "created_at": "2025-02-05T12:49:16.182445827Z",
      "id": "972e129e48bcf3420113f0d8d1570d5c3dbb5045ad46cc688c692c76e7cf9382",
      "name": "eloquent_cartwright",
      "image": "busybox:latest",
      "labels": {

      },
      "running": true,
      "ports": [],
      "status": "running",
      "volumes": {
        "testing": "/testing"
      }
    }
  }
]
```


Sample error response:

```
{
  "message": "Internal error fetching containers.",
  "detail": "GET http://[fd7a:115c:a1e0:4ca7:9e0e:688f:a0d8:ac55]:4/api/v0/containers: unexpected status code 500: Could not get containers.\n\tError: get containers: run docker ps: exit status 1"
}
```

johnstcn committed:
feat(agent): add container list handler

Adds an API endpoint to coderd `/api/v2/workspaceagents/:id/containers`
that allows listing containers visible to the agent. This initial
implementation only supports listing containers using the Docker CLI.
Support for other data sources may be added at a future date.

johnstcn commented on coderd/util/maps/maps.go:
```diff
@@ -0,0 +1,14 @@
+
+// Subset returns true if all the keys of a are present
+// in b and have the same values.
+func Subset[T, U comparable](a, b map[T]U) bool {
+	for ka, va := range a {
+		if vb, ok := b[ka]; !ok {
+			return false
+		} else if va != vb {
+			return false
+		}
+	}
+	return true
+}
```
review: I couldn't find a stdlib function for this, so rolled my own. I'd love to get rid of this.

johnstcn commented on agent/containers_dockercli.go:
```diff
@@ -0,0 +1,191 @@
+		return nil, xerrors.Errorf("run docker inspect: %w", err)
+	}
+
+	// out := make([]codersdk.WorkspaceAgentContainer, 0)
```
```suggestion
```

johnstcn commented on agent/containers_dockercli.go:
```diff
@@ -0,0 +1,191 @@
+		return nil, xerrors.Errorf("run docker ps: %w", err)
+	}
+
+	// the output is returned with a single item per line, so we have to decode it
+	// line-by-line
```
More applicable for `docker ps --all --no-trunc --format json`.

```suggestion
```

johnstcn commented on agent/containers_dockercli.go:
```diff
@@ -0,0 +1,191 @@
+
+// convertDockerPort converts a Docker port string to a port number and network
+// example: "8080/tcp" -> 8080, "tcp"
+//
```
```suggestion
```

johnstcn commented on agent/containers_dockercli.go:
```diff
@@ -0,0 +1,191 @@
+// container path. If the host path is not specified, the container path is used
+// as the host path.
+// example: "/host/path=/container/path" -> "/host/path", "/container/path"
+//
```
```suggestion
```

johnstcn commented on agent/containers_dockercli.go:
```diff
@@ -0,0 +1,191 @@
+	return out
+}
+
+// convertDockerPort converts a Docker port string to a port number and network
```
If there's pre-existing code to handle this I'd be happy to use that instead.

johnstcn commented on agent/containers_dockercli.go:
```diff
@@ -0,0 +1,191 @@
+	}
+}
+
+// convertDockerVolume converts a Docker volume string to a host path and
```
Again, if there's pre-existing code I'd be happy to use this.

johnstcn commented on codersdk/workspaceagents.go:
```diff
@@ -392,6 +392,68 @@ func (c *Client) WorkspaceAgentListeningPorts(ctx context.Context, agentID uuid.
+	// Running is true if the container is currently running.
+	Running bool `json:"running"`
+	// Ports includes ports exposed by the container.
+	Ports []WorkspaceAgentListeningPort `json:"ports"`
```
I'm re-using this struct, as any exposed port should also function the same way as a "listening port".

johnstcn commented on codersdk/workspaceagents.go:
```diff
@@ -392,6 +392,68 @@ func (c *Client) WorkspaceAgentListeningPorts(ctx context.Context, agentID uuid.
 
+// WorkspaceAgentContainer describes a container of some sort that is visible
+// to the workspace agent.
+type WorkspaceAgentContainer struct {
```
Open to calling this anything else, "Container" has a lot of associated baggage, whereas these are basically ephemeral resources created at runtime that we want to surface to users. Could be a Docker container, could be a Kubernetes pod, could be a `chroot`.

dannykopping replied:
I suspect you'll need different types for each, otherwise this abstraction is gonna be bloody bloody
(`chroot` has no concept of image / ports).

mafredri replied:
I feel this could be called `WorkspaceAgentDevcontainer`, and thus be the abstraction layer between different implementations. I'd prefer we output one standardized type so we don't need to handle them in various places (UI, CLI, etc).

johnstcn committed:
Apply suggestions from code review

johnstcn committed:
lint

dannykopping commented on agent/containers.go:
```diff
@@ -0,0 +1,91 @@
+func (ch *containersHandler) handler(rw http.ResponseWriter, r *http.Request) {
+	ct, err := ch.getContainers(r.Context())
+	if err != nil {
+		httpapi.Write(r.Context(), rw, http.StatusInternalServerError, codersdk.Response{
```
For later: we should consider adding more specific error handling here
i.e. if the docker CLI can't be located, permissions error, etc
We could then send back specific codes to the calling API to provide a clearer error explanation & links to docs for resolution

dannykopping commented on agent/containers.go:
```diff
@@ -0,0 +1,91 @@
+	ch.containers = updated
+	ch.mtime = now
+
+	// return a copy
```
Nit: I find comments like these redundant; it's clear a copy is being returned
A more valuable comment could be the "why"...
I assume it's for encapsulation & concurrency safety purposes

dannykopping commented on agent/containers_dockercli.go:
```diff
@@ -0,0 +1,188 @@
+func (*dockerCLIContainerLister) List(ctx context.Context) ([]codersdk.WorkspaceAgentContainer, error) {
+	var buf bytes.Buffer
+	// List all container IDs, one per line, with no truncation
+	cmd := exec.CommandContext(ctx, "docker", "ps", "--all", "--quiet", "--no-trunc")
```
Does this work across all versions of docker?

dannykopping replied:
Thanks for going back a few versions to check 👍 

I think we can leave it for now. If we find this causes issues later we can refactor the args into version types.

dannykopping commented on agent/containers_dockercli.go:
```diff
@@ -0,0 +1,188 @@
+	cmd := exec.CommandContext(ctx, "docker", "ps", "--all", "--quiet", "--no-trunc")
+	cmd.Stdout = &buf
+	if err := cmd.Run(); err != nil {
+		return nil, xerrors.Errorf("run docker ps: %w", err)
```
We're leaving stderr on the floor here (I think); would it not be useful to send that back, or is that already included in `err`?

johnstcn replied:
Yeah, I'll add `stderr` to the error here:

```
run docker ps: exit status 1: "permission denied while trying to connect to the Docker daemon socket at unix:///var/run/docker.sock: Get \"http://%2Fvar%2Frun%2Fdocker.sock/v1.47/containers/json?all=1\": dial unix /var/run/docker.sock: connect: permission denied"
```

dannykopping commented on agent/containers_dockercli.go:
```diff
@@ -0,0 +1,188 @@
+		return nil, xerrors.Errorf("run docker ps: %w", err)
+	}
+
+	ids := make([]string, 0)
+	for _, line := range strings.Split(buf.String(), "\n") {
```
Nit: we can preallocate here, assuming all lines are relevant
Most lines in `docker ps` output are relevant.

```suggestion
        lines := strings.Split(buf.String(), "\n")
	ids := make([]string, 0, len(lines))
	for _, line := range lines {
```

dannykopping commented on agent/containers_dockercli.go:
```diff
@@ -0,0 +1,188 @@
+		return nil, xerrors.Errorf("run docker inspect: %w", err)
+	}
+
+	ins := make([]dockerInspect, 0)
+	if err := json.NewDecoder(&buf).Decode(&ins); err != nil {
+		return nil, xerrors.Errorf("decode docker inspect output: %w", err)
+	}
+
+	out := make([]codersdk.WorkspaceAgentContainer, 0)
```
Ditto for these

dannykopping commented on agent/containers_dockercli.go:
```diff
@@ -0,0 +1,188 @@
+		ID:           in.ID,
+		Image:        in.Config.Image,
+		Labels:       in.Config.Labels,
+		Ports:        make([]codersdk.WorkspaceAgentListeningPort, 0),
```
```suggestion
		Ports:        make([]codersdk.WorkspaceAgentListeningPort, 0, len(in.Config.ExposedPorts)),
```

dannykopping commented on agent/containers_dockercli.go:
```diff
@@ -0,0 +1,188 @@
+		Ports:        make([]codersdk.WorkspaceAgentListeningPort, 0),
+		Running:      in.State.Running,
+		Status:       in.State.String(),
+		Volumes:      make(map[string]string),
```
```suggestion
		Volumes:      make(map[string]string, len(in.Config.Volumes)),
```

dannykopping commented on agent/containers_internal_test.go:
```diff
@@ -0,0 +1,206 @@
+	ct, err := pool.RunWithOptions(&dockertest.RunOptions{
+		Repository: "busybox",
+		Tag:        "latest",
+		Cmd:        []string{"sleep", "infnity"},
```
Ha! TIL that `infblah` is a valid arg to `sleep` since it only scans for an `inf` suffix

Anyway I'm sure this was a typo, so for correctness' sake:

```suggestion
		Cmd:        []string{"sleep", "infinity"},
```

dannykopping commented on agent/containers_internal_test.go:
```diff
@@ -0,0 +1,206 @@
+	})
+	require.NoError(t, err, "Could not start test docker container")
+	t.Cleanup(func() {
+		assert.NoError(t, pool.Purge(ct), "Could not purge resource")
```
Might be useful to print the container name if this fails so we can manually clean it up

dannykopping commented on agent/containers_internal_test.go:
```diff
@@ -0,0 +1,206 @@
+				expected: []codersdk.WorkspaceAgentContainer{fakeCt2},
+			},
+		} {
+			tc := tc
```
Nit: no longer necessary as of Go v1.21.

johnstcn replied:
I know, my linter just complains.

dannykopping commented on coderd/workspaceagents.go:
```diff
@@ -678,6 +679,92 @@ func (api *API) workspaceAgentListeningPorts(rw http.ResponseWriter, r *http.Req
+	if !ok {
+		labelParam = []string{}
+	}
+	labels := make(map[string]string)
```
```suggestion
	labels := make(map[string]string, len(labelParam)/2)
```

dannykopping commented on coderd/workspaceagents.go:
```diff
@@ -678,6 +679,92 @@ func (api *API) workspaceAgentListeningPorts(rw http.ResponseWriter, r *http.Req
+
+	// If the agent is unreachable, the request will hang. Assume that if we
+	// don't get a response after 30s that the agent is unreachable.
+	ctx, cancel := context.WithTimeout(ctx, 30*time.Second)
```
For later: let's add some specific error handling here; if the request times out, we shouldn't let this context cancelation bubble up through the http stack but rather return a specific `408 Request Timeout` (unless it does this already and I'm an ignoramus).

johnstcn replied:
Added below (L754)

dannykopping replied:
This? https://github.com/coder/coder/blob/d4aa89bd9211358375864ba69dbb465ab03b0913/coderd/workspaceagents.go#L754
I must be missing something

johnstcn replied:
Might be looking at the old code. I added a 408 if error is context.Canceled.

dannykopping commented on coderd/workspaceagents.go:
```diff
@@ -678,6 +679,92 @@ func (api *API) workspaceAgentListeningPorts(rw http.ResponseWriter, r *http.Req
+		return
+	}
+
+	// Filter in-place by labels
```
Curious why we don't pass the label filtering down to where the data is collected?
[`docker` can filter by label](https://docs.docker.com/reference/cli/docker/container/ls/#filter)

johnstcn replied:
I found that co-locating filtering and caching gets difficult.

mafredri replied:
That's a good point Cian, although I'd question the premise of caching at this stage, did you find it was a requirement to have any form of responsive UI? Is `docker ps` too slow?

It would be easier to just run multiple `docker ps` commands in parallel if there are multiple requests inbound.

johnstcn replied:
Eh, perhaps. The caching logic isn't overly complex, and I was mainly copying the pattern from the `listening-ports` handler. I can remove it if you feel strongly about it, but I suspect it'll be needed -- especially if someone is running lots of containers in their workspace.

dannykopping commented on coderd/workspaceagents_test.go:
```diff
@@ -1053,6 +1055,87 @@ func TestWorkspaceAgentListeningPorts(t *testing.T) {
 	})
 }
 
+func TestWorkspaceAgentContainers(t *testing.T) {
```
Vibe: seems a bit redundant to have this test in both layers
I assume this is because it's easier than starting up an agent and a tailnet to test the full integration?

johnstcn replied:
It does actually start an agent and tailnet! This tests the end-to-end flow of client -> coderd -> agent -> docker ps -> agent -> coderd -> client.

dannykopping replied:
Aha!

dannykopping reviewed (commented):
Great work 👍 
Left mostly small comments

mafredri commented on coderd/workspaceagents_test.go:
```diff
@@ -1053,6 +1055,87 @@ func TestWorkspaceAgentListeningPorts(t *testing.T) {
+		t.Skip("this test creates containers, which is flaky on non-linux runners")
+	}
+
+	pool, err := dockertest.NewPool("")
```
I'd prefer to use a fake container here if we need it, or perhaps a fake API response from the agent, instead of introducing docker dependency to coderd tests.

johnstcn replied:
I kept the existing test but hid it behind an environment variable. Added a separate test and plumbed through a mock implementation.

mafredri commented on coderd/workspaceagents_test.go:
```diff
@@ -1053,6 +1055,87 @@ func TestWorkspaceAgentListeningPorts(t *testing.T) {
+	ct, err := pool.RunWithOptions(&dockertest.RunOptions{
+		Repository: "busybox",
+		Tag:        "latest",
+		Cmd:        []string{"sleep", "infnity"},
```
Nity!

mafredri commented on codersdk/workspaceagents.go:
```diff
@@ -392,6 +392,68 @@ func (c *Client) WorkspaceAgentListeningPorts(ctx context.Context, agentID uuid.
+
+// WorkspaceAgentContainersLabelFilter is a RequestOption for filtering
+// listing containers by labels.
+func WorkspaceAgentContainersLabelFilter(kvs map[string]string) RequestOption {
```
This seems like it could be a private function? Let's not cloud the API package with functions that aren't useful for the consumer.

mafredri reviewed (commented):
Nice work, Cian! Looking past nits and suggestions, in broad strokes I'd like to see:

- Separate agent/containers package
- No explicit dependency on docker for testing (e.g. skip)
- Single "container" output from agent that can be consumed in a generic way in e.g. UI, so that docker can be swapped out for another backend and everything behaves the same
- `s/container/devcontainer/` makes sense to me, but perhaps I'm overlooking something

johnstcn commented on agent/containers.go:
```diff
@@ -0,0 +1,91 @@
+package agent
```
Yeah I was just starting to think that. 👍

johnstcn commented on agent/containers.go:
```diff
@@ -0,0 +1,91 @@
+
+func (ch *containersHandler) getContainers(ctx context.Context) ([]codersdk.WorkspaceAgentContainer, error) {
+	ch.mu.Lock()
+	defer ch.mu.Unlock()
```
TIL about the "channel-as-lock" pattern 👍

johnstcn commented on agent/containers_dockercli.go:
```diff
@@ -0,0 +1,188 @@
+			return 0, "", xerrors.Errorf("invalid port format: %s", in)
+		}
+		return uint16(p), parts[1], nil
+	}
```
Added test cases.

johnstcn commented on agent/containers_internal_test.go:
```diff
@@ -0,0 +1,206 @@
+func fakeContainer(t *testing.T, mut ...func(*codersdk.WorkspaceAgentContainer)) codersdk.WorkspaceAgentContainer {
+	t.Helper()
+	ct := codersdk.WorkspaceAgentContainer{
+		CreatedAt:    time.Now().UTC(),
```
I'm not sure the juice is worth the squeeze; the caller can override via `mut`.

johnstcn commented on agent/containers_dockercli.go:
```diff
@@ -0,0 +1,188 @@
+		return in, in
+	case 1:
+		return parts[0], parts[0]
+	default:
```
In order to properly expose this to the user, I'll introduce a `Warnings []string` field to the API response.

johnstcn committed:
address code review comments

johnstcn committed:
move to agentcontainers package

johnstcn committed:
remove docker dependency from tests

johnstcn committed:
update .gitattributes

johnstcn commented on agent/agentcontainers/containers_dockercli.go:
```diff
@@ -0,0 +1,214 @@
+	stderrBuf.Reset()
+	// nolint: gosec // We are not executing user input, these IDs come from
+	// `docker ps`.
+	cmd = exec.CommandContext(ctx, "docker", append([]string{"inspect"}, ids...)...)
```
Review: Linter says I need to use `agentexec.Execer` instead here.

johnstcn commented on agent/agentcontainers/containers_dockercli.go:
```diff
@@ -0,0 +1,214 @@
+func (*DockerCLILister) List(ctx context.Context) (codersdk.WorkspaceAgentListContainersResponse, error) {
+	var stdoutBuf, stderrBuf bytes.Buffer
+	// List all container IDs, one per line, with no truncation
+	cmd := exec.CommandContext(ctx, "docker", "ps", "--all", "--quiet", "--no-trunc")
```
Apparently I need to use `agentexec.Execer` here.

mafredri replied:
Oh yeah, I forgot we had this package now. It's good we have linting for this, although in this instance I would almost argue `agentexec` isn't _strictly_ necessary as the agent is exec'ing Docker as an extension of its functionality.

**But** let's use `agentexec` anyway. At worst it might result in our exec of Docker not being granted high enough priority to finish its work in time, at best it protects us against Docker CLI going haywire and using too much memory/cpu, or something like that.

johnstcn commented:
@dannykopping @mafredri I think I'm ready for round 2 now. I still need to use `agentexec.Execer` instead but I think I've addressed all of your comments.

mafredri commented on agent/agentcontainers/containers_dockercli.go:
```diff
@@ -0,0 +1,214 @@
+	// List all container IDs, one per line, with no truncation
+	cmd := exec.CommandContext(ctx, "docker", "ps", "--all", "--quiet", "--no-trunc")
+	cmd.Stdout = &stdoutBuf
+	cmd.Stderr = &stderrBuf
```
Would it be worth logging any output to stderr even if the command succeeds? Either agent logging or as a warning entry in API response?

mafredri reviewed (approved):
This PR is looking quite good! Just a few minor things and while testing this PR I realized we don't include mounts in the output. I think that could be relevant, like volumes, but I'll defer to you whether or not we include those.

We should still make the `agentexec` changes, but other than that, approved!

johnstcn committed:
handle r.Context() canceled

johnstcn committed:
move mock to its own package

johnstcn committed:
init lockCh in New(), remove sync.Once

johnstcn committed:
use agentexec

dannykopping reviewed (approved):
LGTM :shipit:

johnstcn committed:
correctly detect ports and volumes

johnstcn committed:
ci: add go tools to build-dylib
